File: examples/dekita/.claude/hooks/issue-body-requirements-check.py
#!/usr/bin/env python3
"""gh issue create時にIssue本文の必須項目をチェックする。

Why:
    調査なしにIssueを作成すると、実装時に問題の把握に時間がかかる。
    必須セクション（なぜ/現状/期待動作）を強制することで、
    Issue作成時点での十分な調査を促す。

What:
    - gh issue createコマンドを検出
    - --body オプションから本文を抽出
    - 必須セクション（なぜ/現状/期待動作）の存在を確認
    - 不足している場合はブロック

Remarks:
    - trivial/documentationラベルでスキップ可能
    - issue-investigation-reminderを置き換えた後継フック

Changelog:
    - silenvx/dekita#2455: フック追加
"""

import json
import os.path
import re
import shlex
import sys
from pathlib import Path

from lib.execution import log_hook_execution
from lib.results import make_approve_result, make_block_result
from lib.session import parse_hook_input

HOOK_NAME = "issue-body-requirements-check"

# 必須セクションのパターン定義
# 各パターンは (名前, 正規表現, 説明) のタプル
REQUIRED_SECTIONS = [
    (
        "なぜ/背景",
        r"^(?:##|###)\s*(?:なぜ|背景|理由|Why|Motivation|Background|Reason)",
        "変更の動機・背景を記載してください",
    ),
    (
        "現状/実際の動作",
        r"^(?:##|###)\s*(?:現状|実際|現在|Current|Actual|Status)",
        "現在の状態・実際の動作を記載してください",
    ),
    (
        "期待動作/対応案",
        r"^(?:##|###)\s*(?:期待|対応|Expected|Proposed|Solution|何を|What)",
        "期待する動作または対応案を記載してください",
    ),
]

# スキップ対象のラベル
SKIP_LABELS = ["trivial", "documentation", "docs"]


def _is_gh_command(token: str) -> bool:
    """Check if a token represents the gh command (bare name or full path)."""
    return os.path.basename(token) == "gh"


def _skip_env_prefixes(parts: list[str]) -> list[str]:
    """Skip VAR=value environment variable prefixes from token list."""
    cmd_start = 0
    for i, token in enumerate(parts):
        if "=" in token and not token.startswith("-"):
            cmd_start = i + 1
        else:
            break
    return parts[cmd_start:]


def is_gh_issue_create_command(command: str) -> bool:
    """Check if command starts with gh issue create."""
    try:
        tokens = shlex.split(command)
    except ValueError:
        return False

    remaining = _skip_env_prefixes(tokens)

    if len(remaining) < 3:
        return False
    return _is_gh_command(remaining[0]) and remaining[1] == "issue" and remaining[2] == "create"


def extract_body_from_command(command: str) -> str | None:
    """Extract --body or --body-file option value from gh issue create command.

    Supports:
    - --body "content" / --body="content" / -b "content"
    - --body-file "path" / --body-file="path" / -F "path" (reads file content)

    If both --body and --body-file are specified, --body takes precedence.
    """
    try:
        tokens = shlex.split(command)
    except ValueError:
        return None

    body: str | None = None
    body_file: str | None = None

    for i, token in enumerate(tokens):
        # --body "value" or --body value
        if token == "--body" and i + 1 < len(tokens):
            if body is None:
                body = tokens[i + 1]
            continue
        # --body="value"
        if token.startswith("--body="):
            if body is None:
                body = token[7:]
            continue
        # -b "value" or -b value (short form)
        if token == "-b" and i + 1 < len(tokens):
            if body is None:
                body = tokens[i + 1]
            continue
        # --body-file "path" or --body-file path
        if token == "--body-file" and i + 1 < len(tokens):
            if body_file is None:
                body_file = tokens[i + 1]
            continue
        # --body-file="path"
        if token.startswith("--body-file="):
            if body_file is None:
                body_file = token[12:]
            continue
        # -F "path" or -F path (short form for --body-file)
        if token == "-F" and i + 1 < len(tokens):
            if body_file is None:
                body_file = tokens[i + 1]
            continue

    # --body takes precedence over --body-file
    if body is not None:
        return body

    # If body-file was specified, read the file content with path traversal protection
    if body_file:
        try:
            # Path traversal protection: only allow files within cwd
            safe_directory = Path.cwd().resolve()
            file_path = Path(body_file).resolve()

            if not str(file_path).startswith(str(safe_directory)):
                # Path is outside the safe directory, reject it
                return None

            if file_path.exists():
                return file_path.read_text(encoding="utf-8")
        except (OSError, UnicodeDecodeError):
            # If file cannot be read, return None (will trigger block)
            pass

    return None


def extract_labels_from_command(command: str) -> list[str]:
    """Extract --label option values from gh issue create command."""
    try:
        tokens = shlex.split(command)
    except ValueError:
        return []

    labels = []
    for i, token in enumerate(tokens):
        if token == "--label" and i + 1 < len(tokens):
            # --label can contain comma-separated values
            labels.extend(tokens[i + 1].split(","))
        elif token.startswith("--label="):
            labels.extend(token[8:].split(","))
        elif token == "-l" and i + 1 < len(tokens):
            labels.extend(tokens[i + 1].split(","))

    return [label.strip().lower() for label in labels]


def should_skip_check(command: str, body: str | None) -> bool:
    """Check if the requirements check should be skipped."""
    # Skip if body contains "調査不要"
    if body and "調査不要" in body:
        return True

    # Skip if trivial or documentation label is present
    labels = extract_labels_from_command(command)
    return any(label in SKIP_LABELS for label in labels)


def check_required_sections(body: str) -> list[tuple[str, str]]:
    """Check if body contains all required sections.

    Returns:
        List of (section_name, description) for missing sections.
    """
    missing = []
    for name, pattern, description in REQUIRED_SECTIONS:
        if not re.search(pattern, body, re.MULTILINE | re.IGNORECASE):
            missing.append((name, description))
    return missing


def main():
    """PreToolUse hook for Bash commands."""
    try:
        data = parse_hook_input()
        tool_input = data.get("tool_input", {})
        command = tool_input.get("command", "")

        # gh issue create コマンドを検出
        if not is_gh_issue_create_command(command):
            sys.exit(0)

        body = extract_body_from_command(command)

        # スキップ条件をチェック
        if should_skip_check(command, body):
            result = make_approve_result(
                HOOK_NAME, "スキップ条件に該当（trivial/documentation/調査不要）"
            )
            log_hook_execution(HOOK_NAME, "approve", "skip condition matched")
            print(json.dumps(result))
            sys.exit(0)

        # bodyがない場合はブロック
        if not body:
            message = "\n".join(
                [
                    "🚫 Issue本文（--body）が指定されていません。",
                    "",
                    "以下のセクションを含めてください:",
                    "- ## なぜ（変更の動機・背景）",
                    "- ## 現状（現在の状態・実際の動作）",
                    "- ## 期待動作 または ## 対応案",
                    "",
                    "スキップするには --label trivial または --label documentation を付与",
                ]
            )
            result = make_block_result(HOOK_NAME, message)
            log_hook_execution(HOOK_NAME, "block", "no body specified")
            print(json.dumps(result))
            sys.exit(2)

        # 必須セクションをチェック
        missing = check_required_sections(body)
        if missing:
            missing_list = "\n".join([f"- {name}: {desc}" for name, desc in missing])
            message = "\n".join(
                [
                    "🚫 Issue本文に必須セクションがありません。",
                    "",
                    "不足しているセクション:",
                    missing_list,
                    "",
                    "スキップするには --label trivial または --label documentation を付与",
                ]
            )
            result = make_block_result(HOOK_NAME, message)
            log_hook_execution(
                HOOK_NAME, "block", f"missing sections: {[name for name, _ in missing]}"
            )
            print(json.dumps(result))
            sys.exit(2)

        # 全て揃っている場合は承認
        result = make_approve_result(HOOK_NAME, "Issue本文の必須項目を確認しました")
        log_hook_execution(HOOK_NAME, "approve", "all required sections present")
        print(json.dumps(result))
        sys.exit(0)

    except Exception as e:
        print(f"[{HOOK_NAME}] Hook error: {e}", file=sys.stderr)
        result = make_approve_result(HOOK_NAME, f"Hook error: {e}")
        log_hook_execution(HOOK_NAME, "approve", f"Hook error: {e}")
        print(json.dumps(result))
        sys.exit(0)


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/issue_body_requirements_check.ts
#!/usr/bin/env bun
/**
 * gh issue create時にIssue本文の必須項目をチェックする。
 *
 * Why:
 *   Creating issues without investigation causes delays during implementation.
 *   Enforce required sections (Why/What/How) to encourage proper investigation
 *   at issue creation time.
 *
 * What:
 *   - Detect gh issue create commands
 *   - Extract body from --body option
 *   - Check for required sections (Why/What/How)
 *   - Block if missing
 *
 * Remarks:
 *   - trivial/documentationラベルでスキップ可能
 *   - issue-investigation-reminderを置き換えた後継フック
 *   - Python版: issue_body_requirements_check.py
 *
 * Changelog:
 *   - silenvx/dekita#2455: フック追加
 *   - silenvx/dekita#2917: TypeScript版初期実装
 *   - silenvx/dekita#3052: 汎用オプションパーサー導入、-F=形式対応
 */

import { existsSync, readFileSync } from "node:fs";
import { basename, relative, resolve } from "node:path";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import {
  type OptionDef,
  getOptionValue,
  getOptionValues,
  parseOptions,
  skipEnvPrefixes,
  tokenize,
} from "../lib/option_parser";
import { makeApproveResult, makeBlockResult } from "../lib/results";
import { parseHookInput } from "../lib/session";

const HOOK_NAME = "issue-body-requirements-check";

// Required sections patterns: [name, pattern, description]
const REQUIRED_SECTIONS: Array<[string, RegExp, string]> = [
  [
    "Why",
    /^(?:##|###)\s*(?:Why|Motivation|Background|Reason)/im,
    "Describe the motivation/background for this issue",
  ],
  [
    "What",
    /^(?:##|###)\s*(?:What|Current|Actual|Status)/im,
    "Describe the current state or what this issue is about",
  ],
  ["How", /^(?:##|###)\s*(?:How|Expected|Proposed|Solution)/im, "Describe the proposed solution"],
];

// Labels that skip the check
const SKIP_LABELS = ["trivial", "documentation", "docs"];

// オプション定義（宣言的に管理）
const OPTION_DEFS: OptionDef[] = [
  { long: "body", short: "b", hasValue: true },
  { long: "body-file", short: "F", hasValue: true },
  { long: "label", short: "l", hasValue: true, multiple: true },
];

/**
 * Check if a token represents the gh command (bare name or full path).
 */
function isGhCommand(token: string): boolean {
  return basename(token) === "gh";
}

/**
 * Check if command is gh issue create.
 */
function isGhIssueCreateCommand(command: string): boolean {
  const tokens = tokenize(command);
  if (tokens.length === 0) {
    return false;
  }

  const remaining = skipEnvPrefixes(tokens);

  if (remaining.length < 3) {
    return false;
  }
  return isGhCommand(remaining[0]) && remaining[1] === "issue" && remaining[2] === "create";
}

/**
 * Extract --body or --body-file option value from gh issue create command.
 */
function extractBodyFromCommand(command: string): string | null {
  const tokens = tokenize(command);
  if (tokens.length === 0) {
    return null;
  }

  const options = parseOptions(tokens, OPTION_DEFS);

  // --body takes precedence over --body-file
  const body = getOptionValue(options, "body");
  if (body !== null) {
    return body;
  }

  // If body-file was specified, read the file content with path traversal protection
  const bodyFile = getOptionValue(options, "body-file");
  if (bodyFile) {
    return readBodyFile(bodyFile);
  }

  return null;
}

/**
 * Read body content from file with path traversal protection.
 */
function readBodyFile(bodyFile: string): string | null {
  try {
    const safeDirectory = resolve(process.cwd());
    const filePath = resolve(bodyFile);

    // Path traversal protection: only allow files within cwd
    const rel = relative(safeDirectory, filePath);
    if (rel.startsWith("..") || rel.startsWith("/")) {
      // Path is outside the safe directory, reject it
      return null;
    }

    if (existsSync(filePath)) {
      return readFileSync(filePath, "utf-8");
    }
  } catch {
    // If file cannot be read, return null (will trigger block)
  }
  return null;
}

/**
 * Extract --label option values from gh issue create command.
 */
function extractLabelsFromCommand(command: string): string[] {
  const tokens = tokenize(command);
  if (tokens.length === 0) {
    return [];
  }

  const options = parseOptions(tokens, OPTION_DEFS);
  const labels = getOptionValues(options, "label");

  // Split comma-separated labels and normalize to lowercase
  return labels.flatMap((label) => label.split(",")).map((label) => label.trim().toLowerCase());
}

/**
 * Check if the requirements check should be skipped.
 */
function shouldSkipCheck(command: string): boolean {
  // Skip if trivial or documentation label is present
  const labels = extractLabelsFromCommand(command);
  return labels.some((label) => SKIP_LABELS.includes(label));
}

/**
 * Check if body contains all required sections.
 */
function checkRequiredSections(body: string): Array<{ name: string; description: string }> {
  const missing: Array<{ name: string; description: string }> = [];
  for (const [name, pattern, description] of REQUIRED_SECTIONS) {
    if (!pattern.test(body)) {
      missing.push({ name, description });
    }
  }
  return missing;
}

async function main(): Promise<void> {
  let sessionId: string | undefined;
  try {
    const data = await parseHookInput();
    sessionId = data.session_id;
    const toolInput = (data.tool_input as Record<string, unknown>) || {};
    const command = (toolInput.command as string) || "";

    // Only check gh issue create commands
    if (!isGhIssueCreateCommand(command)) {
      process.exit(0);
    }

    // Check skip conditions
    if (shouldSkipCheck(command)) {
      const result = makeApproveResult(HOOK_NAME, "Skip condition matched (trivial/documentation)");
      await logHookExecution(HOOK_NAME, "approve", "skip condition matched", undefined, {
        sessionId,
      });
      console.log(JSON.stringify(result));
      process.exit(0);
    }

    const body = extractBodyFromCommand(command);

    // Block if no body
    if (!body) {
      const message = [
        "🚫 Issue body (--body) is not specified.",
        "",
        "Please include the following sections:",
        "- ## Why (motivation/background)",
        "- ## What (current state)",
        "- ## How (proposed solution)",
        "",
        "To skip, add --label trivial or --label documentation",
      ].join("\n");
      const result = makeBlockResult(HOOK_NAME, message);
      await logHookExecution(HOOK_NAME, "block", "no body specified", undefined, { sessionId });
      console.log(JSON.stringify(result));
      process.exit(2);
    }

    // Check required sections
    const missing = checkRequiredSections(body);
    if (missing.length > 0) {
      const missingList = missing
        .map(({ name, description }) => `- ${name}: ${description}`)
        .join("\n");
      const message = [
        "🚫 Issue body is missing required sections.",
        "",
        "Missing sections:",
        missingList,
        "",
        "To skip, add --label trivial or --label documentation",
      ].join("\n");
      const result = makeBlockResult(HOOK_NAME, message);
      await logHookExecution(
        HOOK_NAME,
        "block",
        `missing sections: ${JSON.stringify(missing.map((m) => m.name))}`,
        undefined,
        { sessionId },
      );
      console.log(JSON.stringify(result));
      process.exit(2);
    }

    // All required sections present
    const result = makeApproveResult(HOOK_NAME, "Issue body requirements verified");
    await logHookExecution(HOOK_NAME, "approve", "all required sections present", undefined, {
      sessionId,
    });
    console.log(JSON.stringify(result));
    process.exit(0);
  } catch (error) {
    console.error(`[${HOOK_NAME}] Hook error: ${formatError(error)}`);
    const result = makeApproveResult(HOOK_NAME, `Hook error: ${formatError(error)}`);
    await logHookExecution(HOOK_NAME, "approve", `Hook error: ${formatError(error)}`, undefined, {
      sessionId,
    });
    console.log(JSON.stringify(result));
    process.exit(0);
  }
}

// Only run main when executed directly, not when imported
if (import.meta.main) {
  main();
}

// Export for testing
export {
  isGhIssueCreateCommand,
  extractBodyFromCommand,
  extractLabelsFromCommand,
  shouldSkipCheck,
  checkRequiredSections,
  readBodyFile, // テスト用にエクスポート（内部ヘルパー）
};
